micronota/bfillings/diamond.py

```diff
@@ -176,85 +179,141 @@ def make_db(in_fp, out_fp=None, params=None):
-                            evalue=0.01, cores=1, params=None):
-    '''Search query sequences against the database.
-
-    Parameters
-    ----------
-    query : str
-        The file path to the query sequence.
-    db : str
-        The file path to diamond formatted database.
-    cores : int
-        Number of CPU cores. Default to 1. If it is set to 0, it will use
-        all available cores.
-    evalue : float
-        Default to 0.01. Threshold E-value.
-    params : dict
-        Other command line parameters for diamond blastp. key is the option
-        (e.g. "-T") and value is the value for the option (e.g. "50").
-        If the option is a flag, set the value to None.
-    Returns
-    -------
-    str
-        The file path of the blast result.
+class FeatureAnnt(MetadataPred):
     '''
-    prefix = basename(query)
-    tmpd = mkdtemp(suffix='', prefix='diamond_', dir=out_dir)
-    daa_fp = join(out_dir, '%s.daa' % prefix)
-    if aligner == 'blastp':
-        app = DiamondBlastp
-    elif aligner == 'blastx':
-        app = DiamondBlastx
-    else:
-        raise ValueError('unknown aliger')
-
-    blast = app(InputHandler='_input_as_paths', params=params)
-    blast.Parameters['--query'].on(query)
-    blast.Parameters['--db'].on(db)
-    blast.Parameters['--evalue'].on(evalue)
-    blast.Parameters['--threads'].on(cores)
-    blast.Parameters['--tmpdir'].on(tmpd)
-    blast.Parameters['--daa'].on(daa_fp)
-    blast_res = blast()
-    blast_res.cleanUp()
-
-    out_fp = join(out_dir, '%s.diamond' % prefix)
-    view = DiamondView(InputHandler='_input_as_paths', params=params)
-    view.Parameters['--daa'].on(daa_fp)
-    view.Parameters['--out'].on(out_fp)
-    view.Parameters['--outfmt'].on(outfmt)
-    view_res = view()
-    view_res.cleanUp()
-    # print(app.BaseCommand)
-    return out_fp
-
-
-def parse_output(diamond_res, column='bitscore'):
-    '''Parse the output of diamond blastp/blastx.
-
-    Parameters
+    Attributes
     ----------
-    diamond_res : str
-        file path
-
-    Returns
-    -------
-    pandas.DataFrame
-        The best matched records for each query sequence.
+    dat : list of str
+        list of file path to databases.
     '''
-    columns = ['qseqid', 'sseqid', 'pident', 'length', 'mismatch',
-               'gapopen', 'qstart', 'qend', 'sstart', 'send',
-               'evalue', 'bitscore']
-    df = pd.read_table(diamond_res, names=columns)
-    # pick the rows that have highest bitscore for each qseqid
-    # df_max = df.groupby('qseqid').apply(
-    #     lambda r: r[r[column] == r[column].max()])
-    idx = df.groupby('qseqid')[column].idxmax()
-    df_max = df.loc[idx]
-    df_max.index = idx.index
-    return df_max[['sseqid', 'evalue', 'bitscore']]
+    def __init__(self, dat, out_dir, tmp_dir=None, cache=None):
+        super().__init__(dat, out_dir, tmp_dir)
+        if cache is not None:
+            self.dat = cache + self.dat
```

Comment:
I think this needs some further discussion.

I think it may be a better idea to have the cache as an instance variable.  Any thoughts?